src/coreclr/jit/morph.cpp

```diff
@@ -12387,6 +12402,72 @@ GenTree* Compiler::fgOptimizeHWIntrinsic(GenTreeHWIntrinsic* node)
 
 #endif
 
+//------------------------------------------------------------------------
+// fgOptimizeDivision: Optimizes division operations.
+//
+// Arguments:
+//   tree - the unchecked GT_DIV/GT_UDIV tree to optimize.
+//
+// Return Value:
+//   The optimized tree that can have any shape.
+//
+GenTree* Compiler::fgOptimizeDivision(GenTreeOp* div)
+{
+    assert(div->OperIs(GT_DIV, GT_UDIV));
+
+    if (gtTreeHasSideEffects(div, GTF_SIDE_EFFECT))
+    {
+        return nullptr;
+    }
+
+    GenTree* op1 = div->gtGetOp1();
+    GenTree* op2 = div->gtGetOp2();
+
+    // Folds cascade of DIVs/UDIVs into one DIV/UDIV and a multiplication
+    // of root and child values
+    if (op1->OperIs(GT_UDIV, GT_DIV) && op2->IsIntegralConst() && op1->gtGetOp2()->IsIntegralConst())
+    {
+
+        GenTree* chOp1 = op1->gtGetOp1();
+        GenTree* chOp2 = op1->gtGetOp2();
+
+        IntegralRange boundRange = IntegralRange::ForNode(op2, this);
+        int64_t       upperBound = IntegralRange::SymbolicToRealValue(boundRange.GetUpperBound());
+
+        int64_t child_val = chOp2->AsIntConCommon()->IntegralValue();
+        int64_t root_val  = op2->AsIntConCommon()->IntegralValue();
+
+        // Make sure it is not folded into overflow
+        // If it is going to overflow, then result of such a division
+        // is 0
+        if ((upperBound / child_val) < root_val)
+        {
+            GenTree* ret = gtNewZeroConNode(TYP_INT);
+            fgUpdateConstTreeValueNumber(ret);
+
+            DEBUG_DESTROY_NODE(div);
+
+            INDEBUG(ret->gtDebugFlags |= GTF_DEBUG_NODE_MORPHED);
+
+            return ret;
+        }
+
+        div->gtOp1 = chOp1;
+        div->gtOp1->SetVNsFromNode(op1);
+
+        div->gtOp2 = chOp2;
+        div->gtOp2->AsIntConCommon()->SetIntegralValue(root_val * child_val);
+        div->gtOp2->SetVNsFromNode(op2);
```

Comment:
this looks like it sets VN that does not correspond to the new constant value

Reply:
Also, since your opt is guarded with fgGlobalMorph you don't need to care about it all I guess (or enable it for all phases)